File: packages/ios_platform_images/ios/Classes/IosPlatformImagesPlugin.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "IosPlatformImagesPlugin.h"

#if !__has_feature(objc_arc)
#error ARC must be enabled!
#endif

@interface IosPlatformImagesPlugin ()
@end

@implementation IosPlatformImagesPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FPIPlatformImagesApiSetup(registrar.messenger, [[IosPlatformImagesPlugin alloc] init]);
}

- (nullable FPIPlatformImageData *)
    loadImageWithName:(nonnull NSString *)name
                error:(FlutterError *_Nullable __autoreleasing *_Nonnull)error {
  UIImage *image = [UIImage imageNamed:name];
  NSData *data = UIImagePNGRepresentation(image);
  if (!data) {
    return nil;
  }
  return [FPIPlatformImageData makeWithData:[FlutterStandardTypedData typedDataWithBytes:data]
                                      scale:@(image.scale)];
}

- (nullable NSString *)resolveURLForResource:(nonnull NSString *)name
                               withExtension:(nullable NSString *)extension
                                       error:(FlutterError *_Nullable __autoreleasing *_Nonnull)
                                                 error {
  NSURL *url = [[NSBundle mainBundle] URLForResource:name withExtension:extension];
  return url.absoluteString;
}

@end

File: packages/ios_platform_images/ios/Classes/IosPlatformImagesPlugin.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Flutter
import Foundation

public final class IosPlatformImagesPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "plugins.flutter.io/ios_platform_images",
      binaryMessenger: registrar.messenger())

    let instance = IosPlatformImagesPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "loadImage":
      loadImage(call, result)
    case "resolveURL":
      resolveURL(call, result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func loadImage(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let name = call.arguments as? String else {
      result(nil)
      return
    }

    guard let image = UIImage(named: name),
      let data = image.pngData()
    else {
      result(nil)
      return
    }

    let imageResult: [String: Any] = [
      "scale": image.scale,
      "data": FlutterStandardTypedData(bytes: data),
    ]

    result(imageResult)
  }

    private func resolveURL(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String?] else {
      return result(nil)
    }

    let name = args[0]
    let extesnsionName = args.count > 1 && args[1] != nil ? args[1] : nil

    guard let url = Bundle.main.url(forResource: name, withExtension: extesnsionName) else {
      return result(nil)
    }
      
    result(url)

  }

}
